### PR TITLE
xds: Adds support for xdstp in State-of-the-World

### PR DESCRIPTION
**Pull request by abeyad**
Previously, xdstp support was limited to the Delta protocol. This commit adds support for xdstp in State-of-the-World (SotW) xDS, for both the unified gRPC mux and the legacy gRPC mux.

Signed-off-by: Ali Beyad <abeyad@google.com>

### CODE REVIEW

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:2393]: Is this a copy-paste of the Basic test?
If so, and assuming that more tests will be added for both of SotW and Delta, I suggest merging the two (test-parameter) and skipping the LEDS tests in SotW mode.

File: source/common/config/subscription_factory_impl.cc
Comment (adisuissa) [inline source/common/config/subscription_factory_impl.cc:151]: I'm probably missing something here, but why is unified mux needed here?

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:1752]: nit: instead of passing `this` consider passing `sotwOrDelta()`.

**Review by adisuissa — COMMENTED** [review]
Thanks, left a few comments.
/wait

File: source/common/config/subscription_factory_impl.cc
Comment (adisuissa) [reply source/common/config/subscription_factory_impl.cc:151]: Not sure I fully grok the strange behavior. If the cluster was updated, it should apply it IMHO.

If it doesn't require many changes then supporting non-unified might be beneficial, as one can easily change the flag and see that everything works.

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:1913]: question: can this be done even before `initialize()` is called?

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:1884]: Not sure I understand this... for LDS (wildcard) removal can be done by sending the rest of the listeners.

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:1853]: I'm not sure why node is being sent. I think you can set `ads_config->set_set_node_on_first_message_only(true);` in the initialization phase.

File: test/integration/ads_integration_test.cc
Comment (adisuissa) [inline test/integration/ads_integration_test.cc:1867]: To keep consistency between SotW and Delta, I believe the 2nd argument needs to pass all the relevant listeners.

**Review by adisuissa — COMMENTED** [review]
Thanks! Left a few more comments.
/wait

**Review by abeyad — COMMENTED** [review]
I implemented xdstp sotw on the legacy mux as well.  PTAL

File: source/common/config/grpc_mux_impl.cc
Comment (adisuissa) [inline source/common/config/grpc_mux_impl.cc:394]: I'm not sure I fully understand why `added_resources` is needed.
It seems that both wildcard and non-wildcard watchers will need to have the resource as part of the `found_resources`, no?

nit: in the TODO I think it would be better to have a dedicated map of wildcard watches, so the search can be inverted (go over the resources, and find the wildcard-watches that is relevant). WDYT?

**Review by adisuissa — COMMENTED** [review]
Thanks, overall LGTM mostly nits.

File: source/common/config/grpc_mux_impl.cc
Comment (abeyad) [reply source/common/config/grpc_mux_impl.cc]: The reason for added_resources is because if we have a watch on a wildcard, we loop through the resource_ref_map, so we could end up adding the same entry from the resource_ref_map more than once.  E.g. if we are watching `xdstp://test/foo` and `xdstp://test/*`, and the resource_ref_map contains `xdstp://test/foo`.  However, instead of having the set keep track of whether we've already inserted it, I just did an erase from the resource_ref_map once we've inserted into found_resources.

RE the TODO, sorry I think I wasn't clear, what I meant is basically what you're suggesting.  I tried clearing up the language.b

Comment (adisuissa) [reply source/common/config/grpc_mux_impl.cc]: I guess my comment wasn't clear. In the case of watching `xdstp://test/foo` and `xdstp://test/*` why aren't both watchers notified with the resource? I guess it could also raise another question: in the current implementation why does singleton-watcher take precedence over collection-watcher?

Comment (abeyad) [reply source/common/config/grpc_mux_impl.cc]: Discussed offline, removed the deduping check as we should not have singleton and collection resources in the same watch; and I commented the assumption.

File: source/common/config/grpc_mux_impl.cc
Comment (adisuissa) [inline source/common/config/grpc_mux_impl.cc:397]: nit: even if we there are a singleton and collection watchers, I think it is ok, because of the `else if` above.
Maybe remove or rephrase it as follows:
```suggestion
        // Note: although it is unlikely that Envoy will need to support a resource that is mapped to both a singleton and collection watch,
        // this code still supports this use case.
```

**Review by adisuissa — COMMENTED** [review]
Thanks! Left one more small nit.

**Review by adisuissa — DISMISSED** [review]
LGTM, thanks!

**Comment by abeyad** [comment]
Thanks @adisuissa !  Adding @lizan for cross-company review.

File: source/common/config/grpc_mux_impl.cc
Comment (lizan) [inline source/common/config/grpc_mux_impl.cc:39]: Nit: XdsTp? Seems other place (hasXdsTpScheme) is using this instead of Xdstp.

File: source/common/config/grpc_mux_impl.cc
Comment (lizan) [inline source/common/config/grpc_mux_impl.cc:50]: Nit: absl::StrCat instead of +

**Review by lizan — COMMENTED** [review]
LGTM modulo two bits. Defer to @adisuissa for final merge.

File: source/common/config/grpc_mux_impl.cc
Comment (abeyad) [reply source/common/config/grpc_mux_impl.cc]: done

File: source/common/config/grpc_mux_impl.cc
Comment (abeyad) [reply source/common/config/grpc_mux_impl.cc]: done

**Comment by abeyad** [comment]
Thanks for the review, @lizan !  @adisuissa can you kindly re-approve since I pushed up a commit to address Lizan's feedback, thanks!

**Review by adisuissa — APPROVED** [review]
LGTM thanks.
@lizan approved as well, so this can be merged.

**Pull request merged**
